### PR TITLE
Add tests to verify format of csv files

### LINKED ISSUE #191: File format consistency tests
I was trying to think of some simple unit tests to verify the formats of the csv files.  However, it's not clear to me what the expected format should be, if there is one.

For example, I have found that files whose names do not end in `precinct.csv` all appear to have the following header:
```
county,office,district,party,candidate,votes
```
Files whose names end with `precinct.csv` have a variety of differences.  Most have the following header:
```
county,precinct,office,district,party,candidate,votes
```
However, there are many that do not. Some differences include:
1. Extra fields after `county,precinct,office,district,party,candidate,votes`.
2. The fields `county,precinct,office,district,party,candidate,votes` appearing in a different order.
3. An "empty" first field (`,county,precinct,office,district,party,candidate,votes`), whose row values are just the row number.
4. An "empty" last field (`county,precinct,office,district,party,candidate,votes,`) whose row values are all empty.
5. Inconsistent casing of headers: `County,Precinct,Office,District,Party,Candidate,Votes,election_day,mail`
6. The presence of a byte order mark `\ufeff`.

Do we have an expected file format, and is there any interest in having tests to verify the existing data?

### PR DESCRIPTION
**Pull request by warwickmm**
This adds some simple tests to verify the format of the csv files:
1. The column headers must be a superset of the expected entries.
2. The column headers must not contain an empty entry.
3. Each row must have a consistent number of entries.
4. Check for existence of forbidden characters.

In doing so, we fixed the following issues:

* `2020/20201103__ca__general__lassen__precinct.csv:` Removed an empty column, and removed an extra column that appeared from line 1040 onwards.
* `2018/counties/20180605__ca__primary__mariposa__precinct.csv:` Removed an empty column.
* `2018/counties/20181106__ca__general__san_bernadino__precinct.csv:` Removed an empty column.
* `2020/20201103__ca__general__solano__precinct.csv:` Removed an empty column.
* `2020/20201103__ca__general__santa_cruz__precinct.csv:` Rename `vote` to `votes`.
* `2020/20201103__ca__general__tulare__precinct.csv:` Use lowercase column headers.
* `2020/20200303__ca__primary__siskiyou__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__alpine__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__amador__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__lake__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__mendocino__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__sierra__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__tulare__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__orange__precinct.csv`: Remove `*` characters.

This fixes #191.

### CODE REVIEW

**Comment by warwickmm** [comment]
Without any changes to the existing csv files, the tests reveal the following errors:
```
AssertionError: '' unexpectedly found in {'', 'candidate', 'county', 'district', 'precinct', 'office', 'votes', 'party'} : File 2018/counties/20181106__ca__general__san_bernadino__precinct.csv has an empty column header.
AssertionError: '' unexpectedly found in {'election_day', '', 'candidate', 'county', 'district', 'precinct', 'office', 'votes', 'mail', 'party'} : File 2018/counties/20180605__ca__primary__mariposa__precinct.csv has an empty column header.
AssertionError: False is not true : File 2020/20201103__ca__general__alpine__precinct.csv has header: {'candidate', 'district', '\ufeffcounty', 'precinct', 'office', 'votes', 'party'}, which is missing: {'county'}.
AssertionError: False is not true : File 2020/20200303__ca__primary__siskiyou__precinct.csv has header: {'candidate', 'district', '\ufeffcounty', 'precinct', 'election day', 'provisional', 'office', 'votes', 'early voting', 'mail', 'party'}, which is missing: {'county
AssertionError: '' unexpectedly found in {'', 'candidate', 'county', 'district', 'precinct', 'provisional', 'office', 'votes', 'mail', 'party'} : File 2020/20201103__ca__general__lassen__precinct.csv has an empty column header.
AssertionError: False is not true : File 2020/20201103__ca__general__lake__precinct.csv has header: {'candidate', 'district', '\ufeffcounty', 'precinct', 'office', 'votes', 'party'}, which is missing: {'county'}.
AssertionError: False is not true : File 2020/20201103__ca__general__mendocino__precinct.csv has header: {'candidate', 'district', '\ufeffcounty', 'precinct', 'office', 'votes', 'party'}, which is missing: {'county'}.
AssertionError: False is not true : File 2020/20201103__ca__general__tulare__precinct.csv has header: {'election_day', 'Votes', 'Party', 'Candidate', 'mail', 'Office', 'Precinct', 'District', '\ufeffCounty'}, which is missing: {'candidate', 'county', 'district', 'prec
AssertionError: False is not true : File 2020/20201103__ca__general__amador__precinct.csv has header: {'candidate', 'district', '\ufeffcounty', 'precinct', 'office', 'votes', 'party'}, which is missing: {'county'}.
AssertionError: False is not true : File 2020/20201103__ca__general__santa_cruz__precinct.csv has header: {'election_day', 'candidate', 'county', 'district', 'vote', 'precinct', 'office', 'mail', 'party'}, which is missing: {'votes'}.
AssertionError: False is not true : File 2020/20201103__ca__general__sierra__precinct.csv has header: {'candidate', 'district', '\ufeffcounty', 'precinct', 'office', 'votes', 'party'}, which is missing: {'county'}.
AssertionError: '' unexpectedly found in {'', 'candidate', 'county', 'district', 'precinct', 'office', 'votes', 'party'} : File 2020/20201103__ca__general__solano__precinct.csv has an empty column header.
```

**Comment by warwickmm** [comment]
After fixing a few formatting issues, the test is now revealing that `2020/20201103__ca__general__lassen__precinct.csv` introduces an extra column (containing some percentage value) beginning on line 1040:
```
AssertionError: 9 != 10 : File 2020/20201103__ca__general__lassen__precinct. has header {'county', 'precinct', 'candidate', 'office', 'district', 'party', 'mail', 'votes', 'provisional'}, but row 1040 is ['Lassen', 'Lassen 501', 'President', '', 'DEM', 'JOSEPH R. BIDEN', '47', '11.11%', '47', '0'].
```


https://github.com/openelections/openelections-data-ca/blob/404291ad4f91e36727c735786ab238f01655c0ab/2020/20201103__ca__general__lassen__precinct.csv#L1038-L1042

I'll work on removing this extra column.

File: 2020/20200303__ca__primary__siskiyou__precinct.csv [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿county,precinct,office,district,party,candidate,votes,early voting,election day,provisional,mail
+county,precinct,office,district,party,candidate,votes,early voting,election day,provisional,mail
```
Comment (warwickmm) [inline 2020/20200303__ca__primary__siskiyou__precinct.csv:1]: A byte order mark was removed here.

File: 2020/20201103__ca__general__alpine__precinct.csv [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿county,precinct,office,district,party,candidate,votes
+county,precinct,office,district,party,candidate,votes
```
Comment (warwickmm) [inline 2020/20201103__ca__general__alpine__precinct.csv:1]: A byte order mark was removed here.

File: 2020/20201103__ca__general__amador__precinct.csv [hunk 3]
```diff
@@ -1,4 +1,4 @@
-﻿county,precinct,office,district,party,candidate,votes
+county,precinct,office,district,party,candidate,votes
```
Comment (warwickmm) [inline 2020/20201103__ca__general__amador__precinct.csv:1]: A byte order mark was removed here.

File: 2020/20201103__ca__general__lake__precinct.csv [hunk 4]
```diff
@@ -1,4 +1,4 @@
-﻿county,precinct,office,district,party,candidate,votes
+county,precinct,office,district,party,candidate,votes
```
Comment (warwickmm) [inline 2020/20201103__ca__general__lake__precinct.csv:1]: A byte order mark was removed here.

File: 2020/20201103__ca__general__mendocino__precinct.csv [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿county,precinct,office,district,party,candidate,votes
+county,precinct,office,district,party,candidate,votes
```
Comment (warwickmm) [inline 2020/20201103__ca__general__mendocino__precinct.csv:1]: A byte order mark was removed here.

File: 2020/20201103__ca__general__sierra__precinct.csv [hunk 6]
```diff
@@ -1,4 +1,4 @@
-﻿county,precinct,office,district,party,candidate,votes
+county,precinct,office,district,party,candidate,votes
```
Comment (warwickmm) [inline 2020/20201103__ca__general__sierra__precinct.csv:1]: A byte order mark was removed here.

**Comment by warwickmm** [comment]
The diff is unwieldy, so below is a summary of the changes to the csv files:

* `2020/20201103__ca__general__lassen__precinct.csv:` Removed an empty column, and removed an extra column that appeard from line 1040 onwards.
* `2018/counties/20180605__ca__primary__mariposa__precinct.csv:` Removed an empty column.
* `2018/counties/20181106__ca__general__san_bernadino__precinct.csv:` Removed an empty column.
* `2020/20201103__ca__general__solano__precinct.csv:` Removed an empty column.
* `2020/20201103__ca__general__santa_cruz__precinct.csv:` Rename `vote` to `votes`.
* `2020/20201103__ca__general__tulare__precinct.csv:` Use lowercase column headers.
* `2020/20200303__ca__primary__siskiyou__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__alpine__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__amador__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__lake__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__mendocino__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__sierra__precinct.csv:` Remove byte order mark.
* `2020/20201103__ca__general__tulare__precinct.csv:` Remove byte order mark.

**Comment by carbonphyber** [comment]
My only feedback is that the BOM byte sequence is sometimes useful for certain client programs when using UTF-16/UTF-32.
It might be worth verifying that all text in all CSV files is UTF-8 compatible before removing the BOM characters and that the text files all use the same Endianness after removing it.

Otherwise, I like the idea and it should help with identifying data integrity issues quickly.

**Comment by warwickmm** [comment]
Thanks @carbonphyber. 

The files that I removed the BOM from are encoded us-ascii, so I believe the endianness is not a conern.
```
$ file -bi $(git show --pretty="" --name-only 2280f34)
text/csv; charset=us-ascii
text/csv; charset=us-ascii
text/csv; charset=us-ascii
text/csv; charset=us-ascii
text/csv; charset=us-ascii
text/csv; charset=us-ascii
text/csv; charset=us-ascii
```

The Travis documentation is not so clear to me on the OSS credits.  It [sounds like](https://docs.travis-ci.com/user/billing-faq/#what-if-i-am-building-open-source) credits have to be requested every time you run out, or negotiate for a renewable amount.  Regardless, I also prefer GitHub actions for the additional ease and flexibility.  I can work on creating a workflow for this.

**Comment by dwillis** [comment]
@warwickmm @carbonphyber thanks for this contribution and discussion. I'd also recommend that we remove percentage columns that exist in the data, if possible. And I agree on removing the BOM and having the encoding as UTF-8.

**Comment by warwickmm** [comment]
Revision acf1d41c07f5c7ba0b90e195767a99a345a67bc8 adds a check for a candidate set of forbidden characters.  Please let me know if some characters should be added/removed from this set.

This additional test revealed that the `2020/20201103__ca__general__orange__precinct.csv` file contained some `*` characters next to a candidate's name.  It seems that the [source file](https://github.com/openelections/openelections-sources-ca/blob/200173509058244c71511155d9db517851baf028/2020/general/2020%20Orange%20County,%20CA%20precinct-level%20results.xlsx) uses `*` to denote the winner, and a few instances remained in the csv file.
```
AssertionError: 0 != 1 : File 2020/20201103__ca__general__orange__precinct.csv contains forbidden characters ['*'] in row 26321: ['Orange', '2002', 'State Assembly', '68', 'REP', '*STEVEN "STEVE" CHOI', '67']
```

These were subsequently removed in revision 59c3f6bbd7dba8c7bad065a2110c827ca119d044.

**Comment by warwickmm** [comment]
@dwillis, are the csv files for other states supposed to follow the same general `county,precinct,office,district,party,candidate,votes` format?  I ran the tests on all the `openelections-data-*` state repositories (including DC, but MD appears to be missing), and there are around 2,800 failures.  Some of these are real (e.g, empty or inconsistent number of columns), but others are due to perhaps a different format being used (e.g., missing `county`, `district`, `precinct`, or `total_votes` instead of `votes`, etc.).

When I have time, I'll go through the obvious ones and either create issues or pull requests for them.

**Comment by dwillis** [comment]
@warwickmm Hmm, CSVs in the data repos should follow that basic format, yes, but they also have county-level results in there that would be missing `precinct` columns, too. Thanks for doing this and I'd be interested in seeing the output.

**Comment by warwickmm** [comment]
As an example, it seems that a lot of the files [here](https://github.com/openelections/openelections-data-ak/tree/master/2016) are missing a `county` column.  A quick glance at the failures indicates around:

* 1000 files that are missing a `county` column.
* 125 files that have a row with a different number of columns than the header has.
* 242 files with a byte order mark.
* 301 files with an empty column header.

**Comment by dwillis** [comment]
@warwickmm So with Alaska we have no counties because Alaska has no counties (in the sense that other states do). The others definitely seem worth exploring.

**Comment by warwickmm** [comment]
Thanks, I didn't know that about Alaska.  

Some of the Arizona data appear to have very different headers, for example 

https://github.com/openelections/openelections-data-az/blob/028e08c874110296427a43b40853c26f399a2fea/2014/20141104__az__general__yuma__precinct.csv

Also in that file, some of the rows appear to be more like metadata, rather than results for a candidate: 

```
precinct_id,precinct_name,race_id,race,candidate_id,candidate,count
1201,PRECINCT 001,999999,Race Statistics,999999,Number of Precincts,1,
1201,PRECINCT 001,999999,Race Statistics,999998,Number of Precincts Reporting,1,
```

I also found a few files with some merge conflict markers in them.  :laughing: 

Before we go too crazy discussing individual issues, I'll try and see if there's a way to summarize the potential problems and create individual tickets in the respective repositories.

**Comment by warwickmm** [comment]
Regarding this pull request, the test for forbidden characters passes here, but will fail for some other states. Some state data contain the question being voted on ("Shall the ... ?") and hence contains a `?`.  If this is too restrictive, I can remove the check for forbidden characters.

**Comment by dwillis** [comment]
@warwickmm A couple of things here: first, thanks for all of this work - as you can see, it's super useful. For Arizona, the individual counties used to produce files with their own formats, and for some of the previous years we simply published those. So I'd say AZ is a separate issue from the others in terms of not adhering to the common format.

On the forbidden characters question, I think we want to be liberal in what we accept, so let's leave in the question marks.

**Comment by warwickmm** [comment]
Thanks @dwillis.  This type of data is new to me, so I appreciate your patience and explanations.

I removed the check for forbidden characters since it is too restrictive (there are a lot of failures for various reasons in data for other states).  We can refine this in the future as needed.  Perhaps we can configure a second type of test that simply warns us of these instead of signaling a test failure.

I don't have any further changes planned for this branch, so please let me know if you have any other suggestions or changes you'd like me to make.

**Pull request merged**